server/handler.go

```diff
@@ -53,31 +53,191 @@ type (
 		AuthOptOut bool
 		// AuthZValidator see AuthZValidatorFn
 		AuthZValidator AuthZValidatorFn
+		// AuthZValidatorExtended see AuthZValidatorV2Fn
+		AuthZValidatorExtended AuthZValidatorV2Fn
+		// Label Optional label(name) of the handler
+		Label string
+		// beforeRequestValidate optional function which is given pointers to all request arguments, so they can be combined just before final validation - i.e.
+		// our typical scenarios - request's payload is extended with orgId provided as path parameter. stuffing that into the actual payload may be required for the validation
+		// to pass (i.e. orgId must be supplied and must be uuid type)
+		beforeRequestValidate beforeRequestValidateFn
+		// responseProcessors - optional collection of response processors
+		responseProcessors []ResponseProcessorFn
 	}
 
 	// AuthZValidatorFn a function that takes the authenticated principal and returns whether the principal is authorized.
 	// return true if the user is authorized
 	// return false if the user is NOT authorized and a string indicated the reason.
 	AuthZValidatorFn func(p *iam.ArmoryCloudPrincipal) (string, bool)
 
+	// AuthZValidatorV2Fn a function that takes the authenticated principal and passes current context, so you can make auth
+	// decision based on the additional conditions (i.e. headers, path parameters, etc. )
+	// return true if the user is authorized
+	// return false if the user is NOT authorized
+	AuthZValidatorV2Fn func(ctx context.Context, p *iam.ArmoryCloudPrincipal) (string, bool)
+
+	beforeRequestValidateFn func(ctx context.Context)
+
 	handler[T, U any] struct {
-		config     HandlerConfig
-		handleFunc func(ctx context.Context, request T) (*Response[U], serr.Error)
+		config          HandlerConfig
+		extractArgsFunc extractRequestArgumentsDelegate[T]
+		handleFunc      handleRequestDelegate[T, U]
+	}
+
+	handleRequestDelegate[T, U any]        func(ctx context.Context, request T) (*Response[U], serr.Error)
+	extractRequestArgumentsDelegate[T any] func(ctx context.Context, request *T) (interface{}, serr.Error)
+
+	HandlerArgument interface {
+		Source() ArgumentDataSource
+	}
+
+	ValidatableHandlerArgument interface {
+		HandlerArgument
+		Check() bool
+	}
+
+	ArmoryPrincipalArgument struct {
+		*iam.ArmoryCloudPrincipal
+	}
+
+	voidArgument struct{}
+
+	ArgumentDataSource int
+
+	Handler1Extensions[REQUEST, RESPONSE any] struct {
```

Comment:
You should be able to use a variadic function here rather than creating a struct per number `HandlerArgument`

Reply:
Since it is a generic, you probably can't use variadics. Some thoughts on this are:
- Use an array of args as the last generic (if possible)
- Use a function on that type to add "context" or the list of args afterward

Just curious about the need to make the args part of the generic struct definition or if they should be added through a function or directly to the properties of the struct later.

Reply:
yeah - the fact it is a generic - requires a dedicated type per 1,2 or 3 args combination. the method to register pre-validaiton is specific to the combination of arguments - and is strongly type bound to the argument types  - no need to use some generic lookup in a bag of parameters or array of parameters - just pure, strongly typed method pointer, checked at compile time. 
There is a test case for that functionality provided.